server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -155,6 +159,122 @@ private static void internalParseDocument(MetadataFieldMapper[] metadataFieldsMa
         }
     }
 
+    private void addIgnoredSourceMissingValues(DocumentParserContext context) throws IOException {
+        Collection<IgnoredSourceFieldMapper.NameValue> ignoredFieldsMissingValues = context.getIgnoredFieldsMissingValues();
+        if (ignoredFieldsMissingValues.isEmpty()) {
+            return;
+        }
+
+        try (
+            XContentParser parser = XContentHelper.createParser(
+                parserConfiguration,
+                context.sourceToParse().source(),
+                context.sourceToParse().getXContentType()
+            )
+        ) {
+            DocumentParserContext newContext = new RootDocumentParserContext(
+                context.mappingLookup(),
+                mappingParserContext,
+                context.sourceToParse(),
+                parser
+            );
+            var nameValues = parseDocForMissingValues(newContext, ignoredFieldsMissingValues);
+            for (var nameValue : nameValues) {
+                context.addIgnoredField(nameValue);
+            }
+        }
+    }
+
+    /**
+     * Simplified parsing version for retrieving the source of a given set of fields.
+     */
+    private static List<IgnoredSourceFieldMapper.NameValue> parseDocForMissingValues(
+        DocumentParserContext context,
+        Collection<IgnoredSourceFieldMapper.NameValue> ignoredFieldsMissingValues
+    ) throws IOException {
+        // Maps full name to the corresponding NameValue entry.
+        Map<String, IgnoredSourceFieldMapper.NameValue> fields = new HashMap<>();
```

Comment:
maybe initialize the `fields` map with size based on `ignoredFieldsMissingValues.size()`?